bindings/nodejs/src/functions/uart.cc

```diff
@@ -32,19 +34,63 @@ class SolUART : public JSHandle<SolUART> {
     static const char *jsClassName() { return "SolUART"; }
 };
 
-static void sol_uart_read_callback(void *user_data, struct sol_uart *uart,
-    unsigned char byte_read) {
+static ssize_t sol_uart_on_data_callback(void *user_data,
+    struct sol_uart *uart, const struct sol_buffer *buf) {
     Nan::HandleScope scope;
     sol_uart_data *uart_data = (sol_uart_data *)user_data;
-    Nan::Callback *callback = uart_data->rx_cb;
+    if (!uart_data)
+        return 0;
+
+    Nan::Callback *callback = uart_data->on_data_cb;
     if (!callback)
-        return;
+        return 0;
+
+    struct sol_str_slice slice;
+    slice = sol_buffer_get_slice(buf);
+    Local<Value> buffer =
+        Nan::CopyBuffer(slice.data, slice.len).ToLocalChecked();
 
     Local<Value> arguments[1] = {
-        Nan::New(byte_read)
+        buffer
     };
-
     callback->Call(1, arguments);
+
+    return slice.len;
+}
+
+static void sol_uart_on_feed_done_callback(void *user_data,
+    struct sol_uart *uart, struct sol_blob *blob, int status) {
+    Nan::HandleScope scope;
+    Local<Value> buffer;
+    sol_uart_data *uart_data = (sol_uart_data *)user_data;
+    if (!uart_data)
+        return;
+
+    if (status < 0) {
+        buffer = Nan::Null();
+    } else {
+        struct sol_str_slice slice;
+        slice = sol_str_slice_from_blob(blob);
+        buffer = Nan::CopyBuffer(slice.data, slice.len).ToLocalChecked();
+    }
+
+    Local<Value> arguments[2] = {
+        buffer,
+        Nan::New(status)
```

Comment:
Handling this already in the low level JS bindings.We don't pass the integer to JS users. We use this in the low level bindings to resolve the promise (i.e fulfill/reject with error) based on the status.